compiler/rustc_lint/src/levels.rs

```diff
@@ -225,13 +234,22 @@ impl<'s> LintLevelsBuilder<'s> {
         attrs: &[ast::Attribute],
         store: &LintStore,
         is_crate_node: bool,
+        source_hir_id: Option<HirId>,
     ) -> BuilderPush {
         let mut specs = FxHashMap::default();
         let sess = self.sess;
         let bad_attr = |span| struct_span_err!(sess, span, E0452, "malformed lint attribute input");
-        for attr in attrs {
-            let Some(level) = Level::from_symbol(attr.name_or_empty(), attr.id.as_u32()) else {
-                continue
+        for (attr_index, attr) in attrs.iter().enumerate() {
+            let level = match Level::from_symbol(attr.name_or_empty(), || {
+                LintExpectationId::Unstable(attr.id)
+            }) {
+                None => continue,
+                Some(Level::Expect(unstable_id)) if source_hir_id.is_some() => {
+                    let stable_id =
+                        self.create_stable_id(unstable_id, source_hir_id.unwrap(), attr_index);
+                    Level::Expect(stable_id)
+                }
+                Some(lvl) => lvl,
```

Comment:
I think this would be clearer if we constructed the `LintExpecationId` before calling `Level::from_symbol` and changed that function back to take a plain `LintExceptationId` instead of a closure. Something like

```rust
let expectation_id = if let Some(hir_id) = source_hir_id {
    LintExpecationId::Stable { hir_id, attr_index }
} else { 
    LintExpecatationId::Unstable(attr.id)
};
let Some(level) = Level::from_symbol(attr.name_or_empty(), expectation_id) else {
    continue;
}
```

If we can use the stable id from the get-go, I don't see why we would need to create an unstable id and then register the stable id in the `expecatation_id_map` right after the fact but maybe I'm missing something 🙂

Reply:
The `expecatation_id_map` is needed to map from unstable to stable IDs. If we move it in front of the `match` statement, we would need an additional check to add the unstable to stable mapping if the level is `Expect`. We could also add each mapping to the list, but I would like to avoid it, as this code runs for every attribute. I'm guessing that this match statement is the most efficient, however that's just a guess.

If you have a suggestion where the ID mapping should be added or how it can be changed, I could run a benchmark :upside_down_face: